src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -427,9 +427,19 @@ def __init__(
         self,
         base: BaseCandidate,
         extras: FrozenSet[str],
+        ireq: Optional[InstallRequirement] = None,
     ) -> None:
+        """
+        :param ireq: the InstallRequirement that led to this candidate, if it
+            differs from the base's InstallRequirement. This will often be the
```

Comment:
I feel maybe we should remove the None case and make this always an ireq instance (i.e. if we didn’t do extra manipulation this would just be `base.ireq`) to avoid an unnecessary branching.

It might be useful for this argument (and attribute) to have a different name to indicate it’s only for reporting.

Reply:
Fine by me. I personally prefer to be explicit when reverting to defaults but it's a thin line in this case.

I moved it up so that the attribute is not optional (50cd318ce), but I think it would be cleaner to leave the constructor argument optional for two reasons:
1. caller interface remains simple for the common case
2. `_ireq` is a "private" attribute on `BaseCandidate`. Accessing it from within the same module is one thing, doing so from `factory` feels to me like a violation of the interface. We could of course rename the attribute if we want.

With this context, how would you like me to proceed?